substrate/client/consensus/beefy/src/keystore.rs

```diff
@@ -15,42 +15,59 @@
 
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
+use codec::Decode;
+
+use sp_application_crypto::{key_types::BEEFY as BEEFY_KEY_TYPE, AppCrypto, RuntimeAppPublic};
+#[cfg(feature = "bls-experimental")]
+use sp_core::ecdsa_bls377;
+use sp_core::{ecdsa, keccak_256};
 
-use sp_application_crypto::{key_types::BEEFY as BEEFY_KEY_TYPE, RuntimeAppPublic};
-use sp_core::keccak_256;
 use sp_keystore::KeystorePtr;
+use sp_std::marker::PhantomData;
 
 use log::warn;
 
-use sp_consensus_beefy::{
-	ecdsa_crypto::{Public, Signature},
-	BeefyAuthorityId,
-};
+use sp_consensus_beefy::{AuthorityIdBound, BeefyAuthorityId, BeefySignatureHasher};
+
+#[cfg(feature = "bls-experimental")]
+use sp_consensus_beefy::ecdsa_bls_crypto;
 
 use crate::{error, LOG_TARGET};
 
-/// Hasher used for BEEFY signatures.
-pub(crate) type BeefySignatureHasher = sp_runtime::traits::Keccak256;
+// impl<T: Codec + Debug + Clone + Ord + Sync + Send + AsRef<[u8]> + AppPublic + AppCrypto +
+// RuntimeAppPublic + BeefyAuthorityId<BeefySignatureHasher> >  AuthorityIdBound for T { type
+// Signature = AppCrypto::Signature; }
 
 /// A BEEFY specific keystore implemented as a `Newtype`. This is basically a
 /// wrapper around [`sp_keystore::Keystore`] and allows to customize
 /// common cryptographic functionality.
-pub(crate) struct BeefyKeystore(Option<KeystorePtr>);
-
-impl BeefyKeystore {
+pub(crate) struct BeefyKeystore<AuthorityId: AuthorityIdBound>(
+	Option<KeystorePtr>,
+	PhantomData<fn() -> AuthorityId>,
+)
+where
+	<AuthorityId as RuntimeAppPublic>::Signature: Send + Sync;
+
+impl<AuthorityId: AuthorityIdBound> BeefyKeystore<AuthorityId>
+where
+	<AuthorityId as RuntimeAppPublic>::Signature: Send + Sync,
+{
 	/// Check if the keystore contains a private key for one of the public keys
 	/// contained in `keys`. A public key with a matching private key is known
 	/// as a local authority id.
 	///
 	/// Return the public key for which we also do have a private key. If no
 	/// matching private key is found, `None` will be returned.
-	pub fn authority_id(&self, keys: &[Public]) -> Option<Public> {
+	pub fn authority_id(&self, keys: &[AuthorityId]) -> Option<AuthorityId> {
 		let store = self.0.clone()?;
 
 		// we do check for multiple private keys as a key store sanity check.
-		let public: Vec<Public> = keys
+		let public: Vec<AuthorityId> = keys
 			.iter()
-			.filter(|k| store.has_keys(&[(k.to_raw_vec(), BEEFY_KEY_TYPE)]))
+			.filter(|k| {
+				store
+					.has_keys(&[(<AuthorityId as RuntimeAppPublic>::to_raw_vec(k), BEEFY_KEY_TYPE)])
+			})
```

Comment:
Dumb question for the `public.len() > 1` check below. 
Is having more than 1 validator key bad? Does the (Beefy) implementation support the scenario where one node has more than one validator keys in its keystore

Reply:
I'm not sure I haven't written that part. But it eventually means that we own two public key of the set and because we are going to sign only with one of them, then the other public key remains inactive indefinitly.

Reply:
Yes having multiple validator keys is an node-operator error, and this just helps to warn the operator.